device/device_opts.go

```diff
@@ -0,0 +1,103 @@
+	MaxRetries int
+	// RootCAs, when not nil, sets a custom set of Root CAs to trust against the broker
+	RootCAs *x509.CertPool
+	// UseMqttStore defines whether the SDK will use a store or not for the underlying
+	// MQTT client. It will use the PersistencyDir when it is activated.
+	UseMqttStore bool
+	// PersistencyDir is a known path that will be used by the SDK for MQTT
+	// persistency, for the default database if used, and for crypto storage unless
+	// CryptoDir is specified.
+	// If none of them are specified, the persistency dir can be empty. Otherwise,
+	// an error will be returned upon creation.
+	//
+	// Creating a Device without persistency is not recommended and should be done only
+	// when aware of the implications. Please read all documentation for details.
+	PersistencyDir string
+	// CryptoDir is a known, available path that will be used by the SDK for storing
+	// certificate and private key for connecting to Astarte. When not specified, it
+	// will use the PersistencyDir if specified, or a temporary directory if no persistency
+	// was enabled. Please keep in mind that if a temporary directory is used for crypto
+	// storage, certificates and private keys will always be regenerated at each startup,
+	// which will take a toll on performances and startup times.
+	CryptoDir string
+
+	// UseDefaultDatabase defines whether the SDK should use its default database (SQLite)
+	// for properties and datastream persistency. It will be stored in the specified
+	// PersistencyDir. When set to true, it takes precedence over Database.
+	UseDefaultDatabase bool
+	// Database is a pointer to a valid, open gorm DB to be used for persistency, provided
+	// UseDefaultDatabase is false. Ensure you're passing a dedicated DB which
+	// can be used exclusively by the Astarte SDK to prevent naming conflicts.
+	// If both this variable is nil and UseDefaultDatabase is false, no database will be used.
+	// Not using a database when you can rely on persistency is strongly discouraged - however,
+	// there might be cases in which such a thing is needed, for example if you cannot use CGO
+	// when compiling your executable.
+	Database *gorm.DB
+}
+
+// NewDeviceOptions will create a new DeviceOptions with some
+// sensible default values.
+func NewDeviceOptions() DeviceOptions {
+	return DeviceOptions{
+		UseMqttStore:        true,
+		UseDefaultDatabase:  true,
+		IgnoreSSLErrors:     false,
+		MaxInflightMessages: 100,
+		AutoReconnect:       true,
+		ConnectRetry:        true,
+		MaxRetries:          10,
+	}
+}
+
+func (opts *DeviceOptions) validate() error {
+	switch {
+	case len(opts.PersistencyDir) == 0 && opts.UseMqttStore:
+		return errors.New("cannot enable mqtt store without a persistencyDir")
+	case len(opts.PersistencyDir) == 0 && opts.UseDefaultDatabase:
+		return errors.New("cannot use default database without a persistencyDir")
+	case opts.Database != nil && opts.UseDefaultDatabase:
```

Comment:
I find this interface quite confusing, because if I pass an explicit database I expect that it gets used without me reminding to _disable_ the default database.
IMHO the `UseDefaultDatabase` flag exists just to allow you to not use _any_ database, so I think that it should be actually be `UseDatabase`, defaulting to true, so the normal cases would be:
- Everything default: use the default database
- The user passes an explicit database (no need to also change the `UseDatabase` flag): that database is used
- Explicit `UseDatabase: false`: no database is used

This way the common cases just requires one option to be changed, and `UseDatabase` is clearer in its intent
Of course `UseDatabase: false` + explicit `Database` is still invalid, but now the error path is the hardest one (since it requires changing 2 options explicitly)